extender.go

```diff
@@ -1,40 +1,47 @@
 package lifecycle
 
 import (
-	"errors"
+	"context"
 	"fmt"
+	"io"
 	"os"
 	"path/filepath"
+	"strconv"
+	"strings"
 	"time"
 
 	"github.com/BurntSushi/toml"
-	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
 	"github.com/google/uuid"
+	"golang.org/x/sync/errgroup"
 
 	"github.com/buildpacks/lifecycle/buildpack"
 	"github.com/buildpacks/lifecycle/internal/extend"
+	"github.com/buildpacks/lifecycle/internal/selective"
 	"github.com/buildpacks/lifecycle/launch"
 	"github.com/buildpacks/lifecycle/log"
 )
 
-const buildIDKey = "build_id"
-
 type Extender struct {
-	AppDir            string
-	GeneratedDir      string
-	GroupPath         string
-	ImageRef          string
-	LayersDir         string
-	PlatformDir       string
-	CacheTTL          time.Duration
-	DockerfileApplier DockerfileApplier
-	Extensions        []buildpack.GroupElement
-	Logger            log.Logger
+	AppDir       string // explicitly ignored by the Dockerfile applier, also the Dockefile build context
+	ExtendedDir  string // output directory for extended image layers
+	GeneratedDir string // input Dockerfiles are found here
+	ImageRef     string // the image to extend
+	LayersDir    string // explicitly ignored by the Dockerfile applier
+	PlatformDir  string // explicitly ignored by the Dockerfile applier
+
+	CacheTTL          time.Duration            // a platform input
+	DockerfileApplier DockerfileApplier        // uses kaniko, BuildKit, or other to apply the provided Dockerfile to the provided image
+	Extensions        []buildpack.GroupElement // extensions are ordered from group.toml
 }
 
 //go:generate mockgen -package testmock -destination testmock/dockerfile_applier.go github.com/buildpacks/lifecycle DockerfileApplier
 type DockerfileApplier interface {
-	Apply(workspace string, digest string, dockerfiles []extend.Dockerfile, options extend.Options) error
+	ImageFor(reference string) (v1.Image, error)
+	Apply(dockerfile extend.Dockerfile, toBaseImage v1.Image, withBuildOptions extend.Options, logger log.Logger) (v1.Image, error)
```

Comment:
Pretty significant refactor, now instead of taking a list of Dockerfiles the applier takes a single Dockerfile. This allows us to consolidate all the CNB-spec logic in the `lifecycle` package without it bleeding into `internal/extend/kaniko`.